### PR TITLE
`azure_rm_batchaccount_info`, `azure_rm_batchaccount`Add Managed Identity support to BatchAccount module

### PR DESCRIPTION
**Pull request by nirarg**
##### SUMMARY
Add Managed Identity support to BatchAccount module

Changed module: azure_rm_batchaccount 

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
`azure_rm_batchaccount` & `azure_rm_batchaccount_info`

### CODE REVIEW

File: plugins/module_utils/azure_rm_common_ext.py [hunk 1]
```diff
@@ -213,44 +257,54 @@ def default_compare(self, modifiers, new, old, path, result):
...
-    def update_identities(self, curr_identity):
-        curr_identity = curr_identity or dict()
-        changed = False
-        current_managed_type = curr_identity.get('type', 'None')
-        current_managed_identities = set(curr_identity.get('user_assigned_identities', {}).keys())
-        param_identity = self.module.params.get('identity', {})
-        param_identities = set(param_identity.get('user_assigned_identities', {}).get('id', []))
-        external_identities = param_identities
+    def update_single_managed_identity(self, curr_identity, new_identity):
+        # Converting from single_managed_identity_spec to managed_identity_spec
```
Comment (p3ck) [inline plugins/module_utils/azure_rm_common_ext.py:261]: Suggest the following change
```
+        new_identity = new_identity or dict()
```

File: plugins/module_utils/azure_rm_common_ext.py [hunk 2]
```diff
@@ -213,44 +257,54 @@ def default_compare(self, modifiers, new, old, path, result):
...
+        user_assigned_identity = new_identity.get('user_assigned_identity', None)
+        if user_assigned_identity is not None:
+            new_identity_converted['user_assigned_identities'] = {
+                'id': [user_assigned_identity]
+            }
+        return self.update_managed_identity(curr_identity=curr_identity, new_identity=new_identity_converted, allow_identities_append=False)
 
+    def update_managed_identity(self, curr_identity=None, new_identity=None, allow_identities_append=True):
+        curr_identity = curr_identity or dict()
+        new_identity = new_identity or self.module.params.get('identity', {})
```
Comment (p3ck) [inline plugins/module_utils/azure_rm_common_ext.py]:
```
-        new_identity = new_identity or self.module.params.get('identity', {})
+        new_identity = new_identity or self.module.params.get('identity', {}) or dict()
```

File: plugins/module_utils/azure_rm_common_ext.py [hunk 3]
```diff
@@ -213,44 +257,54 @@ def default_compare(self, modifiers, new, old, path, result):
...
+        curr_managed_type = curr_identity.get('type', 'None')
+        new_managed_type = new_identity.get('type', 'None')
         # If type set to None, and Resource has None, nothing to do
-        if 'None' in param_identity.get('type', 'None') and current_managed_type == 'None':
+        if new_managed_type == 'None' and curr_managed_type == 'None':
             pass
```
Comment (p3ck) [inline plugins/module_utils/azure_rm_common_ext.py]:
```
-            pass
+            return False, None
```

File: plugins/module_utils/azure_rm_common_ext.py [hunk 4]
```diff
@@ -213,44 +257,54 @@ def default_compare(self, modifiers, new, old, path, result):
...
+           'UserAssigned' in new_managed_type and \
+           new_identity.get('user_assigned_identities', {}).get('append', True) is True:
+            result_user_assigned_identities = new_user_assigned_identities.union(curr_user_assigned_identities)
+
+        # Check if module args identities are different as current ones
+        if curr_user_assigned_identities.difference(result_user_assigned_identities) != set():
+            changed = True
 
-        new_identity = self.managed_identity['identity'](type=param_identity.get('type'))
+        result_identity = self.managed_identity['identity'](type=new_identity.get('type'))
```
Comment (p3ck) [inline plugins/module_utils/azure_rm_common_ext.py]:
```
-        result_identity = self.managed_identity['identity'](type=new_identity.get('type'))
+        result_identity = self.managed_identity['identity'](type=new_managed_type)
```

File: plugins/module_utils/azure_rm_common_ext.py [hunk 5]
```diff
@@ -213,44 +257,54 @@ def default_compare(self, modifiers, new, old, path, result):
...
-            new_identity.user_assigned_identities = {}
-            for identity in external_identities:
-                new_identity.user_assigned_identities[identity] = self.managed_identity['user_assigned']()
+        if len(result_user_assigned_identities) > 0:
+            result_identity.user_assigned_identities = {}
+            for identity in result_user_assigned_identities:
+                result_identity.user_assigned_identities[identity] = self.managed_identity['user_assigned']()
 
-        return changed, new_identity
+        return changed, result_identity
```
Comment (p3ck) [inline plugins/module_utils/azure_rm_common_ext.py:314]: For backwards compat
```
+
+    update_identities = update_managed_identity
```

**Review by p3ck — CHANGES_REQUESTED** [review]
Looks good.  Just a couple of small changes.

File: plugins/module_utils/azure_rm_common_ext.py (hunk 1 above)
Comment (nirarg) [reply plugins/module_utils/azure_rm_common_ext.py:261]: Added

File: plugins/module_utils/azure_rm_common_ext.py (hunk 2 above)
Comment (nirarg) [reply plugins/module_utils/azure_rm_common_ext.py]: Fixed

Comment (nirarg) [reply plugins/module_utils/azure_rm_common_ext.py]: I also added comment here to remove `self.module.params.get('identity', {})` after all modules are aligned to this change

File: plugins/module_utils/azure_rm_common_ext.py (hunk 4 above)
Comment (nirarg) [reply plugins/module_utils/azure_rm_common_ext.py]: Changed

File: plugins/module_utils/azure_rm_common_ext.py (hunk 5 above)
Comment (nirarg) [reply plugins/module_utils/azure_rm_common_ext.py:314]: I added this with a comment to remove it after all modules are changed to the new name

File: plugins/module_utils/azure_rm_common_ext.py [hunk 6]
```diff
@@ -213,44 +257,61 @@ def default_compare(self, modifiers, new, old, path, result):
...
-                if current_managed_identities.difference(external_identities) != set():
-                    changed = True
+        if allow_identities_append and \
+           'UserAssigned' in new_managed_type and \
+           new_identity.get('user_assigned_identities', {}).get('append', True) is True:
+            result_user_assigned_identities = new_user_assigned_identities.union(curr_user_assigned_identities)
 
-        new_identity = self.managed_identity['identity'](type=param_identity.get('type'))
+        # Check if module args identities are different as current ones
+        if curr_user_assigned_identities.difference(result_user_assigned_identities) != set():
```
Comment (p3ck) [inline plugins/module_utils/azure_rm_common_ext.py]: Please make the following change:
```
-        if curr_user_assigned_identities.difference(result_user_assigned_identities) != set():
+        if result_user_assigned_identities.difference(curr_user_assigned_identities) != set():
```
The order matters:

```
>>> foo
{'one'}
>>> bar
{'one', 'two'}
>>> foo.difference(bar)
set()
>>> bar.difference(foo)
{'two'}
```

**Review by p3ck — CHANGES_REQUESTED** [review]

File: plugins/module_utils/azure_rm_common_ext.py (hunk 6 above)
Comment (nirarg) [reply plugins/module_utils/azure_rm_common_ext.py]: Fixed, thanks

**Review by p3ck — APPROVED** [review]

**Review by xuzhang3 — APPROVED** [review]

**Pull request merged**